Feat: Language fragment to jetpack compose screen

jackq97 opened the pull request:
Fixes #4160 

- De-coupled and lifted the states of language screen.
- added preview to language screen to better manage the ui.
- added tests.
- updated the animation logic to Launched effects to avoid circular dependency.

jackq97 commented:
@MohitMaliDeveloper  ok, so working again the issues happening with detekt are  LongMethod error at `LanguageScreen` and `AppBarTextField` after decoupling every composable function.

MohitMaliFtechiz commented:
@jackq97 Thanks for making the changes :clap:, I will review this ASAP.

MohitMaliFtechiz commented on app/src/main/java/org/kiwix/kiwixmobile/language/composables/LoadingIndicator.kt:
```diff
@@ -15,13 +15,22 @@
+import androidx.compose.ui.Modifier
+
+@Composable
+fun LoadingIndicator() {
```
@jackq97 I have created a reusable `ContentLoadingProgressBar` in https://github.com/kiwix/kiwix-android/pull/4268/files#diff-e02e2bf321ca88e93589bfabcdc917309398308c5716b7b8d7609efd0ee54407R36 according to our theme, but that is not merged yet. But you can take a reference from there: otherwise I have to change it later.

MohitMaliFtechiz reviewed (changes requested):
@jackq97 Thanks for making the changes, needs a few improvements please see the review changes.

jackq97 commented on app/src/main/java/org/kiwix/kiwixmobile/language/LanguageFragment.kt:
```diff
@@ -18,90 +18,121 @@
+          KiwixAppBar(
+            titleId = R.string.select_languages,
+            navigationIcon = {
+              AppBarNavigationIcon(
```
This was done to reduce the function size as it was causing detekt error . it's hand of hard to compress composable screens . maybe i can think of some workarounds.

MohitMaliFtechiz replied:
You should split the function in smaller parts to avoid detekt errors like I have split in https://github.com/kiwix/kiwix-android/pull/4268/files#diff-b7b43c5c66650b2562423f3e8d96f92541cdd638ace7d6dd1a126b4d4e24bc4dR85

jackq97 commented on app/src/main/java/org/kiwix/kiwixmobile/language/composables/AppBarTextField.kt:
```diff
@@ -0,0 +1,115 @@
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun AppBarTextField(
```
sure!!

jackq97 commented:
@MohitMaliFtechiz  here's the changes

- added inner padding to scaffolding to account for topAppBar.
- moved AppBarTextField to ui/components so it can be re-used.
- shifted hardcoded values to ComposeDimens.
- Added ContentLoadingProgressBar modified it to support
Indeterminate mode
- further decoupling of composable functions

jackq97 commented on app/src/main/java/org/kiwix/kiwixmobile/language/LanguageScreen.kt:
```diff
@@ -0,0 +1,80 @@
+
+  Column(modifier = Modifier.fillMaxSize()) {
+    // spacer to account for top app bar
+    Spacer(modifier = Modifier.height(56.dp))
```
Ok, so the reason why i didn't do it before because the inner padding also accounted for the bottom bar and shifted the whole ui up when we clicked on search bar to input text.

![Screenshot 2025-03-20 161208](https://github.com/user-attachments/assets/e55ee283-1079-4b58-9ccf-f362e9e1386d)

Now, better possible fix is to use compose navigation and passing the inner padding in the root of our navhost so all screens are equally aware of the bottom and top bars.
but in my recent pr what i set up the inner padding for all the sides except the bottom one.

you can test it again and maybe share some feedback

MohitMaliFtechiz replied:
@jackq97 I will do it and let you know.

MohitMaliFtechiz commented:
> moved AppBarTextField to ui/components so it can be re-used.

@jackq97 Have you added the support for the clear button inside it instead of modifying it from the `menuIcon` logic for showing the clear icon?

MohitMaliFtechiz commented:
> > > moved AppBarTextField to ui/components so it can be re-used.
> > 
> > 
> > @jackq97 Have you added the support for the clear button inside it instead of modifying it from the `menuIcon` logic for showing the clear icon?
> 
> Oh yeah good point. I did not we can certainly add an icon at end and put it inside appBarTextField. I will implement it.

Thanks please rename the component as well to define its functionality probably `KiwixSearchView` is a good choice.

jackq97 commented:
@MohitMaliFtechiz  done with the changes ❤️

jackq97 commented on app/src/main/java/org/kiwix/kiwixmobile/language/LanguageScreen.kt:
```diff
@@ -0,0 +1,122 @@
+  actionMenuItemList: List<ActionMenuItem>,
+  onClearClick: () -> Unit,
+  onAppBarValueChange: (String) -> Unit,
+  content: @Composable() () -> Unit,
```
i agree , it is a terrible name but if you have just one composable lambda in your parameter, you get lint issue insisting it should be named "content" but i can suppress it if you want.

MohitMaliFtechiz replied:
@jackq97 We should keep the better and exact naming for variables so that everyone can easily understand its purpose. Please suppress the lint warning, and change the name of this parameter.

jackq97 commented:
@MohitMaliFtechiz here's the requested changes :) 

- added orders to the test rule to avoid retries error in tests.
- renamed content lambda to "navigationIcon".
- removed redundant "Experimental api" from kiwixSearchView.
- added appropriate text size for kiwixSearchView.

jackq97 commented:
> @jackq97 Have you pushed the changes? I am not seeing the new changes.

i did , in the giithub tests i'm seeing "checking for the ability to merge automatically..." . i don't know what caused this

jackq97 commented:
@MohitMaliFtechiz  it was so github error i pushed an empty commit and now the tests are running fine let me know if you see the changes.

MohitMaliFtechiz commented:
@jackq97 Now I can see your changes. Thanks. Please remove the empty commit.

MohitMaliFtechiz commented:
@jackq97 I will do it, here are some conflicts as well in your PR.

jackq97 commented:
> @jackq97 I will do it, here are some conflicts as well in your PR.

i will check and resolve them .

MohitMaliFtechiz commented:
@jackq97 Do this way:

* Update your main branch with kiwix's main branch.
* Rebase you current branch on main.
* After resolving the conflicts, force push your code in this PR.

MohitMaliFtechiz commented:
@jackq97 Thanks, i will have a look at this issue, and place a fix for this.

jackq97 commented:
> @jackq97 Thanks, i will have a look at this issue, and place a fix for this.

sure let me know if i need to make any changes :)

MohitMaliFtechiz reviewed (changes requested):
@jackq97 For the top bar height issue, I am debugging what is wrong with that. In the meantime, can you please do the review changes?

jackq97 commented:
> @jackq97 For the top bar height issue, I am debugging what is wrong with that. In the meantime, can you please do the review changes?

It shifts down the ui when we open keyboard to input text in `TopAppBar`. def some android's inner padding issue and it only happens when we put `TextField` inside the `TopAppBar` title so this is buggy before we used `Row` to make our TopAppBar but since we using proper jetpack compose `TopAppBar`. it's causing this issue.

MohitMaliFtechiz commented:
@jackq97 It is a `windowInsets` issue. `TopAppBar` uses default toolbar insets, which take twice when the searchbar is opened in the `topBar`. Because we have enabled the `edgeToEdge` mode in our application. I am debugging the issue.

jackq97 commented on core/src/main/java/org/kiwix/kiwixmobile/core/ui/components/KiwixSearchView.kt:
```diff
@@ -0,0 +1,90 @@
+          Icon(
+            painter = painterResource(R.drawable.ic_clear_white_24dp),
+            tint = Color.White,
+            contentDescription = null
```
sure I'll add new string in our resource file.

jackq97 commented:
> @jackq97 It is a `windowInsets` issue. `TopAppBar` uses default toolbar insets, which take twice when the searchbar is opened in the `topBar`. Because we have enabled the `edgeToEdge` mode in our application. I am debugging and fixing the issue.

yes i had worked on it before to fix, but didn't want to mess with activity manifest files too much 😅.
thanks for the insights.

MohitMaliFtechiz commented:
@jackq97 I have placed a fix for it. See screenshots.

![Screenshot from 2025-04-02 14-44-36](https://github.com/user-attachments/assets/5a24a217-1bfe-4a18-9dd7-4a761e1cd848)

![Screenshot from 2025-04-02 14-44-43](https://github.com/user-attachments/assets/0ddb806d-6be7-4934-87d3-5a673d027dc6)

Now, please do the review changes so that we can finally merge this PR. Please rebase the PR as well.

jackq97 commented:
> @jackq97 I have placed a fix for it. See screenshots.
> 
> ![Screenshot from 2025-04-02 14-44-36](https://private-user-images.githubusercontent.com/34593983/429416784-5a24a217-1bfe-4a18-9dd7-4a761e1cd848.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1ODY3MzMsIm5iZiI6MTc0MzU4NjQzMywicGF0aCI6Ii8zNDU5Mzk4My80Mjk0MTY3ODQtNWEyNGEyMTctMWJmZS00YTE4LTlkZDctNGE3NjFlMWNkODQ4LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDA5MzM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTczMzlhOWJjNDg3YjZkOTAxNjU1OTE0NDQ4MTFhODQzYTEyZjRiNWRhZGRiNGUzMzAzNjJmNTNjZTg0MTkwNGUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.n6mRh-ohSpy-PKZKJMVOIxPlFiQ0VVx_fPG7bElR-E4)
> 
> ![Screenshot from 2025-04-02 14-44-43](https://private-user-images.githubusercontent.com/34593983/429416813-0ddb806d-6be7-4934-87d3-5a673d027dc6.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1ODY3MzMsIm5iZiI6MTc0MzU4NjQzMywicGF0aCI6Ii8zNDU5Mzk4My80Mjk0MTY4MTMtMGRkYjgwNmQtNmJlNy00OTM0LTg3ZDMtNWE2NzNkMDI3ZGM2LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDA5MzM1M1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWI5OWI4NGQ3NDYxZTZkZmNjOTZiMjIxZTFhNTBiNTJmOGI0ZDM0YzRjNGQ4ZDIwZWRjYWUwMDEzOGM0YjU0ZTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.BwBnRpGhkHx5iNP19pywkUeW09JkXXnoPOEJDDSAaAA)
> 
> Now, please do the review changes so that we can finally merge this PR. Please rebase the PR as well.

i just forced pushed the changes that you requested before i saw this fix.
how can i merge the changes with my current branch and check for changes ?
and eventually i will update the main branch and will rebase it and push it.

also the changes i pushed are 
 used minimumInteractiveComponentSize() on the modifier to avoid "Touch target" accessibility issue.
- added compose dimes to previous missed elements.
- added content description for clear icon button in search view.

MohitMaliFtechiz commented:
> I just forced pushed the changes you requested before I saw this fix.

When you force pushed the code at that time, my fix was gone because it was not in the commit list. No problem, I will repush that.

jackq97 commented:
> > I just forced pushed the changes you requested before I saw this fix.
> 
> When you force pushed the code at that time, my fix was gone because it was not in the commit list. No problem, I will repush that.

yeah i needed to update my branch before pushing it !!. thanks for re pushing now my changes are in too. do i still need to rebase main branch and push the updated code ?

MohitMaliFtechiz commented:
@jackq97 I have rebased and pushed my changes. Now, you don't need to rebase and push the code.

MohitMaliFtechiz reviewed (approved):
@jackq97 Thanks for your hard work :clap: :tada:. @kelson42 LGTM.

jackq97 commented:
> @jackq97 Thanks for your hard work 👏 🎉. @kelson42 LGTM.

thank you for your feedback as well . It was good learning 😊😊😊.
